### PR TITLE
Improve parameter test and implementation

### PR DESCRIPTION
**Pull request by heplesser**
As a step towards developing good test styles, this PR provides a re-write of `test_parameter.py` using PyTest features. It extends the scope of the tests and includes minor fixes:

- float / Parameter is now supported
- when attempting ** with Parameter as exponent, a TypeError is raised to conform with usual Python practice
-

### CODE REVIEW

File: testsuite/pytests/test_parameter.py [hunk 1]
```diff
@@ -20,108 +20,127 @@
...
-
-
-def run():
-    runner = unittest.TextTestRunner(verbosity=2)
-    runner.run(suite())
+This set of tests confirms that arithmetic operations on NEST Parameter objects yield
+correct results by comparing with operations on the undelying numeric values.
+It also confirms that operations on Parameter objects and plain numbers work.
 
+@note These tests only confirm that operators on parameters work in principle. Therefore,
```
Comment (nicolossus) [inline testsuite/pytests/test_parameter.py:29]: Isn't `@note` more of a Doxygen command? I don't mind the usage, just wondering if there is a more appropriate way to highlight notes. For instance, my editor highlights `NOTE` (and also `TODO`) in Python docstrings but not `@note`.

**Review by nicolossus — APPROVED** [review]
LGTM, thanks!

File: testsuite/pytests/test_parameter.py (hunk 1 above)
Comment (jougs) [reply testsuite/pytests/test_parameter.py:29]: I agree. I think you can use proper Restructured Text markup in docstrings, as those also get rendered nicely when we include the tests in the documentation (which we probably should at one point).

Comment (heplesser) [reply testsuite/pytests/test_parameter.py]: Fixed.

File: testsuite/pytests/test_parameter.py [hunk 2]
```diff
@@ -20,108 +20,127 @@
...
+      we can use constant parameters for simplicity. We need to test elsewhere that
```
[the remark this answers is not in the document]
Comment (heplesser) [reply testsuite/pytests/test_parameter.py]: Those tests do not exist yet. I removed the "We need ..." part of the comment and created #2630 instead. We should use issues for things to be done, not hide todos in comments ;).

File: testsuite/pytests/test_parameter_operators.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                           31,
+                           nest.CreateParameter('constant', {'value':  5})],
+                          [operator.mod,
+                           nest.CreateParameter('constant', {'value': 31}),
+                           nest.CreateParameter('constant', {'value':  5})],
+                          [operator.mod,
+                           nest.CreateParameter('constant', {'value':  31}), 5],
+                          [operator.pow,
+                           31,
+                           nest.CreateParameter('constant', {'value':  5})]])
```
Comment (jougs) [inline testsuite/pytests/test_parameter_operators.py:54]:
```suggestion
import operator as op


def const_param(val):
    return nest.CreateParameter('constant', {'value': val})


@pytest.mark.parametrize('op, a, b', [
    [op.pow, const_param(31), const_param(5)],
    [op.pow,             31 , const_param(5)],
    [op.mod, const_param(31), const_param(5)],
    [op.mod, const_param(31),             5 ],
    [op.pow,             31 , const_param(5)]
])
```

**Review by jougs — COMMENTED** [review]
For me, the `parametrize` lines are really hard to read. I have added a suggestion to make them easier.

**Comment by heplesser** [comment]
@jougs I have now applied your suggestion to the remaining tests as well. I changed `import operator as op` to `... as ops` to avoid having the same name for the module with all operands and for the local variable representing the individual operation inside each test.

**Review by jougs — APPROVED** [review]
Thanks! I think this is very nice now.

**Comment by heplesser** [comment]
@jougs Could you resolve the conversations so we can merge?

**Pull request merged**